src/spec.ts

```diff
@@ -355,3 +356,124 @@ export function isStacked(spec: TopLevel<FacetedCompositeUnitSpec>, config?: Con
   }
   return false;
 }
+
+export function selectedFields(spec: NormalizedSpec): string[] {
+  if (isFacetSpec(spec) || isRepeatSpec(spec)) {
+    return selectedFieldsSingle(spec);
+  }
+  if (isLayerSpec(spec)) {
+    return selectedFieldsLayered(spec);
+  }
+  if (isUnitSpec(spec)) {
+    return selectedFieldsUnit(spec);
+  }
+  throw new Error(log.message.INVALID_SPEC);
+}
+
+function selectedFieldsUnit(spec: NormalizedUnitSpec): string[] {
+  const fields: string[] = [];
+  forEach(spec.encoding, (fieldDef, channel) => {
+    fields.push(fieldDef.field);
+  });
+  return fields;
+}
+
+function selectedFieldsLayered(spec: NormalizedLayerSpec): string[] {
+  let fields: string[] = [];
+  spec.layer.map(subspec => {
+    fields = fields.concat(selectedFields(subspec));
+  });
+  return fields;
+}
+
+function selectedFieldsSingle(spec: NormalizedFacetSpec | NormalizedRepeatSpec): string[] {
+  return selectedFields(spec.spec);
+}
+
+export function extractTransforms(spec: NormalizedSpec, config: Config): NormalizedSpec {
```

Comment:
Let's move all this transform extraction code to a new file. Make sure to export the functions from index.ts so that it's available in the API.